src/main/java/hudson/remoting/PingThread.java

```diff
@@ -103,13 +103,13 @@ public void run() {
     private void ping() throws IOException, InterruptedException {
         Future<?> f = channel.callAsync(new Ping());
         long start = System.currentTimeMillis();
-        long end = start +timeout;
 
-        long remaining;
+        long end = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeout);
+        long remaining = end - System.nanoTime();
+
         do {
-            remaining = end-System.currentTimeMillis();
             try {
-                f.get(Math.max(0,remaining),MILLISECONDS);
+                f.get(Math.max(TimeUnit.MILLISECONDS.toNanos(10),remaining),TimeUnit.NANOSECONDS);
```

Comment:
Also what is the logic in choosing 10ms... for GigE the latency between the master and slave should be less than 10ms and all you really want to have happen is that it wakes up at all, it would seem a 1ns minimum wait time might make more sense than 10ms... additionally you probably want this to wake up anyway every 10 seconds or so to allow you to check if the channel has fallen down and bomb out faster, so perhaps you want a `Math.min(TimeUnit.SECONDS.toNanos(10), Math.max(1, remaining))` instead so that it kind-of spin loops